### PR TITLE
Issue285enh semver

### LINKED ISSUE #285: add "Find the latest version satisfying given range" example
Given a set of version strings, find the latest version in a set that satisfies the given range (like docs.rs does in its urls). Use asserts to verify correct serialization and deserailization.

Use the [semver](https://github.com/steveklabnik/semver) crate

cc https://github.com/rust-lang-nursery/rust-cookbook/issues/276

### PR DESCRIPTION
**Pull request by anna-liao**
fixes #285 with additional pre-release example.

r? @budziq @ludwigpacifici 
This is not compiling.  I am stuck on the separate function.  Do you see what might be the issue?  When I run the code in the block comment, it works properly.

### CODE REVIEW

**Comment by ludwigpacifici** [comment]
Hello @anna-liao 

I am a bit confused. Can you tell me what's the relation between this PR and  PR #371? I was expecting to do everything in #371. Let me know if I missed something.

File: src/app.md [hunk 1]
```diff
@@ -725,6 +726,119 @@ fn run() -> Result<()> {
...
+
+use semver::{Version, VersionReq, AlphaNumeric, Numeric};
+#
+# error_chain! {
+#     foreign_links {
+#         SemVer(semver::ReqParseError);
+#     }
+# }
+
+fn version_latest<'a>(
```
Comment (budziq) [inline src/app.md]: I'd probably try to give a more descriptive name here like `find_max_matching_version`

File: src/app.md [hunk 2]
```diff
@@ -725,6 +726,119 @@ fn run() -> Result<()> {
...
+    version_req_str: &str,
+    version_str_vec: &'a Vec<&str>,
```
Comment (budziq) [inline src/app.md]: I'd suggest taking `IntoIter` as a parameter instead of `Vec`

File: src/app.md [hunk 3]
```diff
@@ -725,6 +726,119 @@ fn run() -> Result<()> {
...
+) -> Result<Version, &'a str> {
+    let vreq = VersionReq::parse(version_req_str)?;
+
+    version_str_vec
+        .iter()
+        .filter_map(|s| Version::parse(s).ok())
+        .filter(|s| vreq.matches(s))
+        .max()
+        .ok_or_else(|| "Unable to find version satisfying given range")
```
Comment (budziq) [inline src/app.md]: Then you'd remove the `ok_or_else` part and wrap the returned value in `Ok()`

File: src/app.md [hunk 4]
```diff
@@ -725,6 +726,119 @@ fn run() -> Result<()> {
...
+            pre: vec![],
+            build: vec![],
+        })
+    );
+
+    assert_eq!(
+        version_latest(
+            ">1.2.3-alpha.3",
+            &vec![
+                "1.2.3-alpha.3",
```
Comment (budziq) [inline src/app.md]: lets add some diversity here with some different prerelease tags

File: src/app.md [hunk 5]
```diff
@@ -725,6 +726,119 @@ fn run() -> Result<()> {
...
+            patch: 3,
+            pre: vec![AlphaNumeric(String::from("alpha")), Numeric(10)],
+            build: vec![],
+        })
+    );
+*/
+
+    assert_eq!(
+        version_latest("<= 1.0.0", &vec!["0.9.0", "1.0.0", "1.0.1"]),
+        Ok(Version {
```
Comment (budziq) [inline src/app.md]: error-chain generated `Result`'s tend not to have `PartialEq` implemented so such assertion will not compile but one in which you'd use `?` after the function call will.

**Comment by anna-liao** [comment]
@budziq wow, thanks for the tips for fixing my code!  There's no way I'd figure that out on my own.  Ok, I'm closing this PR and moving back to the original PR.

**Pull request closed without merging**